### ISSUE TITLE
Auto webhook creation broken for Bitbucket 6.x and earlier

### ISSUE THREAD

**Issue by ssc327** [jenkinsci/bitbucket-branch-source-plugin#326]
## Note from project maintainer
> This issue still exists for Bitbucket v5.x.  Fixing that is tracked in #352.
> This has been fixed for Bitbucket v6.x in v2.9.1 of the plugin. 
> To enable the fix, you must go to your Bitbucket Endpoint Configuration and set the "Server Version" to "Bitbucket 6.x (and earlier)".  
> The credentials being used must also have [sufficient permissions to create hooks on the the Bitbucket project](https://github.com/jenkinsci/bitbucket-branch-source-plugin/issues/326#issuecomment-668074806).

## Description
Jenkins version - 2.22.1
Plugin version - 2.8.0
Bitbucket server and version - 5.4.1

When I perform a 'Scan Organization Folder Now', I expect webhooks to be registered for my repos but this does not occur and instead I see 400 errors in the logs.

```
Could not register hooks for sandbox/wordpress-app
com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketRequestException: HTTP request error. Status: 400: .
HttpResponseProxy{HTTP/1.1 400  [X-AREQUESTID: *16U2BIYx857x12710733x0, X-ASEN: SEN-6051868, X-AUSERID: 9658, X-AUSERNAME: username, Cache-Control: no-cache, no-transform, Vary: X-AUSERNAME,Accept-Encoding, Transfer-Encoding: chunked, X-Content-Type-Options: nosniff, Content-Type: application/json;charset=UTF-8, Date: Fri, 10 Jul 2020 18:17:06 GMT, Connection: close] org.apache.http.client.entity.DecompressingEntity@325b29b7}
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.doRequest(BitbucketServerAPIClient.java:1049)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:1002)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:990)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.registerCommitWebHook(BitbucketServerAPIClient.java:600)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHook(WebhookAutoRegisterListener.java:177)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHooks(WebhookAutoRegisterListener.java:133)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.access$000(WebhookAutoRegisterListener.java:56)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener$1.doRun(WebhookAutoRegisterListener.java:102)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:91)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by mtheiss** [comment]
The issue should be the new event for BB 7.x. Older BitBuckets Server do not have it and will fail during webhook creation/update.

pr:from_ref_updated

**Comment by ssc327** [comment]
@mtheiss are you able to point me to the commit which caused the issue?

Is this something that will be fixed or do we need to stop supporting these old server versions?

**Comment by bitwiseman** [comment]
@ssc327 
Please update to plugin version 2.9.0.

**Comment by markush81** [comment]
> @ssc327
> Please update to plugin version 2.9.0.

actually we see same issue with 2.9.0 as well.

**Comment by bitwiseman** [comment]
@mtheiss 
So, because the plugin requests the `pr:from_ref_updated` hook, older bitbucket servers throw this error?

**Comment by bitwiseman** [comment]
@ssc327 
I assume you are running Jenkins "2.222.1" not "2.22.1" right? 

Could you try downgrading to an earlier version of the plugin via http://mirrors.jenkins-ci.org/plugins/cloudbees-bitbucket-branch-source/  ?  Once we know that version where this started happening we can narrow it down. 

As a side note: Bitbucket 5.4.1 is no [longer supported](https://confluence.atlassian.com/support/atlassian-support-end-of-life-policy-201851003.html) by Atlassian.  This doesn't seem like an issue that would be addressed by upgrading to a later 5.x, but it is possible.

**Comment by ssc327** [comment]
@bitwiseman yes, "2.222.1" sorry for the confusion.
I downgraded the plugin to 2.7.0 and then webhook creation worked.

We plan to upgrade to latest BB version later this year but in the interim we are stuck on this old version.

**Comment by markush81** [comment]
I can confirm issue with:

Bitbucket: v6.7.0
Jenkins: 2.235.2 LTS
Plugin: 2.9.0

Issue was not there with 2.7.0.
Currently we have not plan yet, to upgrade Bitbucket.

**Comment by mtheiss** [comment]
> @mtheiss
> So, because the plugin requests the `pr:from_ref_updated` hook, older bitbucket servers throw this error?

@bitwiseman 
That is correct. We have BitBucke v6.9.1 and the issue occurs there. I removed the hook from WebhookConfiguration.java and it works again.

**Comment by mtheiss** [comment]
> 
> 
> @mtheiss are you able to point me to the commit which caused the issue?
> 
> Is this something that will be fixed or do we need to stop supporting these old server versions?

The issue started to appear with commit e35b6c97fbd47651a0910a3e679bcd98602e482a

**Comment by bitwiseman** [comment]
@mtheiss 
Except that that commit is not in 2.8.0, which @ssc327 says also reproduces their problem. 
Are you sure that these are the same issue? 

Still, we probably need to make a high priority change to mitigate that change.

**Comment by lelandsindttouchnet** [comment]
Wanting to verify that this does/does not effect my installation, I upgraded to 2.9.0, setup a new repo and scanned it, the webhooks were created without issue.    ¯\_(ツ)_/¯ 

Jenkins version - 2.235.2
Plugin version - 2.9.0
Bitbucket server and version - 6.4.0

```
Started
[Wed Jul 22 16:58:56 UTC 2020] Starting organization scan...
[Wed Jul 22 16:58:56 UTC 2020] Updating actions...
Looking up team details of SPORK...
Connecting to https://bitbucket.server.name.local/bitbucket using jenkins/****** (Bitbucket Server Endpoint Key)
Team: Spork
[Wed Jul 22 16:58:58 UTC 2020] Consulting Bitbucket Team/Project
Connecting to https://bitbucket.server.name.local/bitbucket using jenkins/****** (Bitbucket Server Endpoint Key)
Looking up repositories of team SPORK
Proposing SPORK/fork
Connecting to https://bitbucket.server.name.local/bitbucket using jenkins/****** (Bitbucket Server Endpoint Key)
Repository type: Git
Looking up SPORK/fork for branches
Checking branch master from SPORK/fork
      ‘Jenkinsfile’ found
    Met criteria

  1 branches were processed (query completed)
Proposing SPORK/spoon
Connecting to https://bitbucket.server.name.local/bitbucket using jenkins/****** (Bitbucket Server Endpoint Key)
Repository type: Git
Looking up SPORK/spoon for branches
Checking branch master from SPORK/spoon
      ‘Jenkinsfile’ found
    Met criteria

  1 branches were processed (query completed)
2 repositories were processed
[Wed Jul 22 16:59:03 UTC 2020] Finished organization scan. Scan took 6.4 sec
Finished: SUCCESS
```

**Comment by markush81** [comment]
> Wanting to verify that this does/does not effect my installation, I upgraded to 2.9.0, setup a new project and scanned it, the webhooks were created without issue. ¯_(ツ)_/¯
> 
> Jenkins version - 2.235.2
> Plugin version - 2.9.0
> Bitbucket server and version - 6.4.0
> 
> ```
> Started
> [Wed Jul 22 16:58:56 UTC 2020] Starting organization scan...
> [Wed Jul 22 16:58:56 UTC 2020] Updating actions...
> Looking up team details of SPORK...
> Connecting to https://bitbucket.server.name.local/bitbucket using jenkins/****** (Bitbucket Server Endpoint Key)
> Team: Spork
> [Wed Jul 22 16:58:58 UTC 2020] Consulting Bitbucket Team/Project
> Connecting to https://bitbucket.server.name.local/bitbucket using jenkins/****** (Bitbucket Server Endpoint Key)
> Looking up repositories of team SPORK
> Proposing SPORK/fork
> Connecting to https://bitbucket.server.name.local/bitbucket using jenkins/****** (Bitbucket Server Endpoint Key)
> Repository type: Git
> Looking up SPORK/fork for branches
> Checking branch master from SPORK/fork
>       ‘Jenkinsfile’ found
>     Met criteria
> 
>   1 branches were processed (query completed)
> Proposing SPORK/spoon
> Connecting to https://bitbucket.server.name.local/bitbucket using jenkins/****** (Bitbucket Server Endpoint Key)
> Repository type: Git
> Looking up SPORK/spoon for branches
> Checking branch master from SPORK/spoon
>       ‘Jenkinsfile’ found
>     Met criteria
> 
>   1 branches were processed (query completed)
> 2 repositories were processed
> [Wed Jul 22 16:59:03 UTC 2020] Finished organization scan. Scan took 6.4 sec
> Finished: SUCCESS
> ```

did you check the jenkins server log? cause this part also looks good for us but in server log we have errors and no hooks are created.

**Comment by lelandsindttouchnet** [comment]
Indeed, the hooks are created.

Jenkins server logs: 

```
2020-07-22 16:58:58.278+0000 [id=52]    INFO    c.c.j.p.b.BitbucketTeamMetadataAction$BitbucketAvatarCacheSource#<init>: Created: BitbucketAvatarSource(https://bitbucket.server.name.local/bitbucket::SPORK::bitbucket-server-endpoint-key)
2020-07-22 16:59:01.226+0000 [id=54]    INFO    c.c.j.p.b.h.WebhookAutoRegisterListener#registerHook: Registering hook for SPORK/fork
2020-07-22 16:59:03.305+0000 [id=52]    INFO    j.b.OrganizationFolder$OrganizationScan#run: SPORK #20200722.165856 organization scan action completed: SUCCESS in 6.4 sec
2020-07-22 16:59:03.408+0000 [id=54]    INFO    c.c.j.p.b.h.WebhookAutoRegisterListener#registerHook: Registering hook for SPORK/spoon
2020-07-22 16:59:03.626+0000 [id=56]    INFO    j.b.MultiBranchProject$BranchIndexing#run: SPORK/fork #20200722.165901 branch indexing action completed: SUCCESS in 2.4 sec
2020-07-22 16:59:05.666+0000 [id=68]    INFO    j.b.MultiBranchProject$BranchIndexing#run: SPORK/spoon #20200722.165903 branch indexing action completed: SUCCESS in 2.3 sec
2020-07-22 16:59:16.293+0000 [id=171]   INFO    o.j.p.workflow.job.WorkflowRun#finish: SPORK/fork/master #1 completed: SUCCESS
2020-07-22 16:59:16.293+0000 [id=170]   INFO    o.j.p.workflow.job.WorkflowRun#finish: SPORK/spoon/master #1 completed: SUCCESS
```

**Comment by markush81** [comment]
> Indeed, the hooks are created.
> 
> Jenkins server logs:
> 
> ```
> 2020-07-22 16:58:58.278+0000 [id=52]    INFO    c.c.j.p.b.BitbucketTeamMetadataAction$BitbucketAvatarCacheSource#<init>: Created: BitbucketAvatarSource(https://bitbucket.server.name.local/bitbucket::SPORK::bitbucket-server-endpoint-key)
> 2020-07-22 16:59:01.226+0000 [id=54]    INFO    c.c.j.p.b.h.WebhookAutoRegisterListener#registerHook: Registering hook for SPORK/fork
> 2020-07-22 16:59:03.305+0000 [id=52]    INFO    j.b.OrganizationFolder$OrganizationScan#run: SPORK #20200722.165856 organization scan action completed: SUCCESS in 6.4 sec
> 2020-07-22 16:59:03.408+0000 [id=54]    INFO    c.c.j.p.b.h.WebhookAutoRegisterListener#registerHook: Registering hook for SPORK/spoon
> 2020-07-22 16:59:03.626+0000 [id=56]    INFO    j.b.MultiBranchProject$BranchIndexing#run: SPORK/fork #20200722.165901 branch indexing action completed: SUCCESS in 2.4 sec
> 2020-07-22 16:59:05.666+0000 [id=68]    INFO    j.b.MultiBranchProject$BranchIndexing#run: SPORK/spoon #20200722.165903 branch indexing action completed: SUCCESS in 2.3 sec
> 2020-07-22 16:59:16.293+0000 [id=171]   INFO    o.j.p.workflow.job.WorkflowRun#finish: SPORK/fork/master #1 completed: SUCCESS
> 2020-07-22 16:59:16.293+0000 [id=170]   INFO    o.j.p.workflow.job.WorkflowRun#finish: SPORK/spoon/master #1 completed: SUCCESS
> ```

interesting, this means it isn't even consistent

**Comment by bitwiseman** [comment]
@lelandsindttouchnet 
Thanks for that information.  

@markush81 @ssc327 @mtheiss 
Hmm, it is unclear what is going on.  Could you provided a bit more of your Jenkins logs on either side of the error for context?  Also, if you could try this with `BitbucketServerAPIClient`  logging set to `FINE`, that might give more information.

**Comment by markush81** [comment]
> @lelandsindttouchnet
> Thanks for that information.
> 
> @markush81 @ssc327 @mtheiss
> Hmm, it is unclear what is going on. Could you provided a bit more of your Jenkins logs on either side of the error for context? Also, if you could try this with `BitbucketServerAPIClient` logging set to `FINE`, that might give more information.

`FINE` does not bring more than that

```
HttpResponseProxy{HTTP/1.1 400  [Server: nginx, Date: Mon, 27 Jul 2020 08:40:59 GMT, Content-Type: application/json;charset=UTF-8, Transfer-Encoding: chunked, X-Cnection: close, X-AREQUESTID: @FVHUMUx520x8335346x0, X-ASEN: SEN-11629201, X-AUSERID: 459, X-AUSERNAME: xxx, Cache-Control: no-cache, no-transform, Vary: accept-encoding,x-auserid,cookie,x-ausername,accept-encoding, X-Content-Type-Options: nosniff, X-Content-Type-Options: nosniff, X-Xss-Protection: 1; mode=block, Strict-Transport-Security: max-age=16070400; includeSubDomains] org.apache.http.client.entity.DecompressingEntity@2a5ce60}
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.doRequest(BitbucketServerAPIClient.java:1072)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:1025)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:1013)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.registerCommitWebHook(BitbucketServerAPIClient.java:623)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHook(WebhookAutoRegisterListener.java:177)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHooks(WebhookAutoRegisterListener.java:133)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.access$000(WebhookAutoRegisterListener.java:56)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener$1.doRun(WebhookAutoRegisterListener.java:102)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:91)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

**Comment by bitwiseman** [comment]
@markush81 
What I want is about 6 lines of logs before and after that exception.  That will help me make sure I'm addressing the issue at hand.

That wire log is useful thanks!

The good news is I believe I have been able to reproduce this locally now using docker images of Bitbucket server 6.x and 7.x.  Woot!

**Comment by bitwiseman** [comment]
@markush81 @mtheiss @ssc327 
I have a PR #332 posted that I believe should address this issue.  
Please install the incremental of the plugin from that PR. 
Then go to you bitbucket server endpoint configuration and change the server version to "Bitbucket v6.x (or later)".  

I have not created automated tests for this yet.  Please report back what you find.  

Thanks!

**Comment by markush81** [comment]
> @markush81 @mtheiss @ssc327
> I have a PR #332 posted that I believe should address this issue.
> Please install [the incremental](https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/cloudbees-bitbucket-branch-source/2.9.1-rc639.c6eb2a3465e9/) of the plugin from that PR.
> Then go to you bitbucket server endpoint configuration and change the server version to "Bitbucket v6.x (or later)".
> 
> I have not created automated tests for this yet. Please report back what you find.
> 
> Thanks!

Does work for me.

**Comment by bitwiseman** [comment]
@markush81 
I'm trying to get it done as soon as possible.

**Comment by bitwiseman** [comment]
@markush81 @mtheiss @ssc327
I've updated the PR.  The new incremental is here: https://repo.jenkins-ci.org/incrementals/org/jenkins-ci/plugins/cloudbees-bitbucket-branch-source/2.9.1-rc644.a528a7df5d64/

@jetersen 
I've written tests for the loading of the endpoint configuration but testing whether it fixes this issue is still manual.  It would be great if we can take merge and release this to unblock people.

**Comment by jetersen** [comment]
@bitwiseman go for a release, we can always fix it in the next version :) Did not mean to try and block anything.

**Comment by bitwiseman** [comment]
@markush81 @mtheiss @ssc327
This has been released as v2.9.1 - please enjoy!

**Comment by ericcitaire** [comment]
I just installed v2.9.1, and I get the following error in Jenkins logs :
```
Could not register hooks for xxxx
com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketRequestException: HTTP request error. Status: 400: .
HttpResponseProxy{HTTP/1.1 400  [X-AREQUESTID: xxxx, X-ASEN: xxxx, X-AUSERID: xxxx, X-AUSERNAME: xxxx, Cache-Control: no-cache, no-transform, Vary: X-AUSERNAME,Accept-Encoding, Transfer-Encoding: chunked, X-Content-Type-Options: nosniff, Content-Type: application/json;charset=UTF-8, Date: Thu, 30 Jul 2020 13:54:07 GMT, Strict-Transport-Security: max-age=15552000; includeSubDomains, Content-Security-Policy: default-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; frame-src https:; img-src 'self' data:, X-Frame-Options: allowall] org.apache.http.client.entity.DecompressingEntity@2313f11e}
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.doRequest(BitbucketServerAPIClient.java:1069)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:1022)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:1010)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.registerCommitWebHook(BitbucketServerAPIClient.java:620)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHook(WebhookAutoRegisterListener.java:177)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHooks(WebhookAutoRegisterListener.java:133)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.access$000(WebhookAutoRegisterListener.java:56)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener$1.doRun(WebhookAutoRegisterListener.java:102)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:91)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

The Webhook is not registered.

I don't have access to BitBucket logs.

**Comment by ericcitaire** [comment]
> I just installed v2.9.1, and I get the following error in Jenkins logs :
> 
> ```
> Could not register hooks for xxxx
> com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketRequestException: HTTP request error. Status: 400: .
> HttpResponseProxy{HTTP/1.1 400  [X-AREQUESTID: xxxx, X-ASEN: xxxx, X-AUSERID: xxxx, X-AUSERNAME: xxxx, Cache-Control: no-cache, no-transform, Vary: X-AUSERNAME,Accept-Encoding, Transfer-Encoding: chunked, X-Content-Type-Options: nosniff, Content-Type: application/json;charset=UTF-8, Date: Thu, 30 Jul 2020 13:54:07 GMT, Strict-Transport-Security: max-age=15552000; includeSubDomains, Content-Security-Policy: default-src 'self' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; script-src 'self' 'unsafe-inline' 'unsafe-eval'; frame-src https:; img-src 'self' data:, X-Frame-Options: allowall] org.apache.http.client.entity.DecompressingEntity@2313f11e}
> 	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.doRequest(BitbucketServerAPIClient.java:1069)
> 	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:1022)
> 	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:1010)
> 	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.registerCommitWebHook(BitbucketServerAPIClient.java:620)
> 	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHook(WebhookAutoRegisterListener.java:177)
> 	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHooks(WebhookAutoRegisterListener.java:133)
> 	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.access$000(WebhookAutoRegisterListener.java:56)
> 	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener$1.doRun(WebhookAutoRegisterListener.java:102)
> 	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:91)
> 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> ```
> 
> The Webhook is not registered.
> 
> I don't have access to BitBucket logs.

Sorry, I didn't configure properly the BitBucket Server Version in Jenkins. It works now.

**Comment by bitwiseman** [comment]
@marx-freedom 
This is issue has been fixed.  Or should be talked about here.

**Comment by bitwiseman** [comment]
IMPORTANT: This has been fixed.  To enable the fix, you must go to your Bitbucket Endpoint Configuration and set the "Server Version" to "Bitbucket 6.x (and earlier)".

**Comment by bitwiseman** [comment]
IMPORTANT: This has been fixed.  To enable the fix, you must go to your Bitbucket Endpoint Configuration and set the "Server Version" to "Bitbucket 6.x (and earlier)".

**Comment by marx-freedom** [comment]
@bitwiseman ok...

I've updated all plugins today, make configuration changes and still have the same problem.

Jenkins 2.235.3	
Bitbucket Branch Source Plugin 2.9.1
Bitbucket Plugin 1.1.16

BitBucket Server 6.10.4

![image](https://user-images.githubusercontent.com/642092/89165183-da1bb300-d580-11ea-97c3-485ab840434e.png)


```
Aug 03, 2020 12:00:09 PM WARNING com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener$1 doRun
Could not register hooks for some-project/some-repo
com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketRequestException: HTTP request error. Status: 401: .
HttpResponseProxy{HTTP/1.1 401  [Server: nginx, Date: Mon, 03 Aug 2020 09:00:09 GMT, Content-Type: application/json;charset=UTF-8, Transfer-Encoding: chunked, Connection: keep-alive, X-AREQUESTID: @1RTONY1x720x29290144x10, X-ASEN: SEN-3190480, X-AUSERID: 182, X-AUSERNAME: jenkins, Cache-Control: no-cache, no-transform, Vary: accept-encoding,x-auserid,cookie,x-ausername,accept-encoding, WWW-Authenticate: OAuth realm="http%3A%2F%2Fsome.host.example%2Fstash", X-Content-Type-Options: nosniff] org.apache.http.client.entity.DecompressingEntity@47908cbe}
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.getRequest(BitbucketServerAPIClient.java:859)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.getWebHooks(BitbucketServerAPIClient.java:708)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHook(WebhookAutoRegisterListener.java:167)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHooks(WebhookAutoRegisterListener.java:133)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.access$000(WebhookAutoRegisterListener.java:56)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener$1.doRun(WebhookAutoRegisterListener.java:102)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:91)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```

Creadentials ok, works for checkout, I can login with it in bitbucket webui without captcha.

**Comment by ssc327** [comment]
@marx-freedom do you always use **plugin** implementation for webhooks? Have you tried native?

**Comment by marx-freedom** [comment]
@ssc327 Yes, I've tried to use native implementation. The Stack trace is slightly different:

```
Aug 03, 2020 4:08:50 PM WARNING com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener$1 doRun
Could not register hooks for some-project/some-repo
com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketRequestException: HTTP request error. Status: 401: .
HttpResponseProxy{HTTP/1.1 401  [Server: nginx, Date: Mon, 03 Aug 2020 13:08:50 GMT, Content-Type: application/json;charset=UTF-8, Transfer-Encoding: chunked, Connection: keep-alive, X-AREQUESTID: @1RTONY1x968x29363729x0, X-ASEN: SEN-3190480, X-AUSERID: 182, X-AUSERNAME: jenkins, Cache-Control: no-cache, no-transform, Vary: accept-encoding,x-auserid,cookie,x-ausername,accept-encoding, WWW-Authenticate: OAuth realm="http%3A%2F%2Fsome.host.example%2Fstash", X-Content-Type-Options: nosniff] org.apache.http.client.entity.DecompressingEntity@56cb1b96}
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.getRequest(BitbucketServerAPIClient.java:859)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.getResources(BitbucketServerAPIClient.java:811)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.getWebHooks(BitbucketServerAPIClient.java:715)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHook(WebhookAutoRegisterListener.java:167)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHooks(WebhookAutoRegisterListener.java:133)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.access$000(WebhookAutoRegisterListener.java:56)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener$1.doRun(WebhookAutoRegisterListener.java:102)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:91)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```

**Comment by ssc327** [comment]
@marx-freedom does the account that you are using have permission to make webhooks on the BB project?

**Comment by marx-freedom** [comment]
@ssc327 thank you for help. 

I've tried different combinations last week, before plugin update, but not all today. 
The permissions setting didn't help last week so I forgot to check this after updating a plugins.
Now I've done new tests using Bitbucket Branch Source Plugin 2.9.1

BitBucket Permissions | WebHook Impl | Hook    | Jenkins log 
----------------------|--------------|---------|------------
write                 | native       | absent  | exception (response code = 401)
write                 | plugin       | absent  | exception (response code = 401)
admin                 | native       | created | "Registering hook for ..."
admin                 | plugin       | created | "Registering hook for ..."

**Comment by bitwiseman** [comment]
@marx-freedom @ssc327 
Thanks for working through that! 
I've added a line to the note at the top of the issue.

**Comment by dsakilesh** [comment]
Hi @bitwiseman 

I have enabled the managehooks and selected bitbucket 6 option. The above solution works for me as well for the newly created jobs. But for the already existing jobs the auto creation of webhook still does not happen.

Should we create the webhook manually for the existing jobs? We have many jobs in place. Please clarify.

Thanks in advance!

**Comment by ssc327** [comment]
@dsakilesh how are you testing webhook creation on existing jobs?

**Comment by dsakilesh** [comment]
Hi @ssc327 

I am using the **Plugin** implementation. I don't see **Jenkins hook** created in Posthooks section of bitbucket repo. And if i create a new branch in the repo then that branch is not visible/trigerred in Jenkins.

**Comment by ssc327** [comment]
@dsakilesh what if you try native? What type of Job are you setting up in Jenkins? Is it a Bitbucket Branch Source job? Have you tried doing an organization scan from the top level job? I have found that is the point where the Hooks are made.

**Comment by dsakilesh** [comment]
Hi @ssc327 
I am using multi branch pipeline branch source job. I did an organisation scan but it did not create the webhook and hence my commits were not picked up by jenkins job. When i delete the job and recreate it then it seems to be creating. But we not want to delete the existing jobs and recreate them. I will try with **Native** implementation and update here.

**Comment by dsakilesh** [comment]
Hi @ssc327 
**Native** implementation does not work at all. New jobs and old jobs both wont create the **Jenkins hook** automatically. Even though i do a Multibranch scan manually they do not create the **Jenkins hook**

**Comment by ssc327** [comment]
@bitwiseman I think that we may need to reopen this issue. I was not able to change the fix previously but I was doing so today and it is still not working. My revised details:

Jenkins: 2.222.1
Plugin: Bitbucket Branch Source Plugin 2.91
BitBucket: Atlassian Bitbucket v5.4.1

I tried a few different config options on the plugin, but the leanest one that should work for sure is below:

![image](https://user-images.githubusercontent.com/39785421/89679082-5fbf9b80-d8be-11ea-831b-402ed6c10a24.png)

The error I get is the following:

```
Could not register hooks for org/app
com.cloudbees.jenkins.plugins.bitbucket.api.BitbucketRequestException: HTTP request error. Status: 400: .
HttpResponseProxy{HTTP/1.1 400  [X-AREQUESTID: *C54G4Fx888x119624587x4, X-ASEN: SEN-6051868, X-AUSERID: 9658, X-AUSERNAME: account, Cache-Control: no-cache, no-transform, Vary: X-AUSERNAME,Accept-Encoding, Transfer-Encoding: chunked, X-Content-Type-Options: nosniff, Content-Type: application/json;charset=UTF-8, Date: Fri, 07 Aug 2020 18:48:24 GMT, Connection: close] org.apache.http.client.entity.DecompressingEntity@4de8e763}
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.doRequest(BitbucketServerAPIClient.java:1069)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:1022)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.postRequest(BitbucketServerAPIClient.java:1010)
	at com.cloudbees.jenkins.plugins.bitbucket.server.client.BitbucketServerAPIClient.registerCommitWebHook(BitbucketServerAPIClient.java:620)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHook(WebhookAutoRegisterListener.java:177)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.registerHooks(WebhookAutoRegisterListener.java:133)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener.access$000(WebhookAutoRegisterListener.java:56)
	at com.cloudbees.jenkins.plugins.bitbucket.hooks.WebhookAutoRegisterListener$1.doRun(WebhookAutoRegisterListener.java:102)
	at hudson.triggers.SafeTimerTask.run(SafeTimerTask.java:91)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
```

Is there perhaps more logging I can enable? I saw that someone had output showing their http requests but I am not sure how to enable that logging.

**Comment by bitwiseman** [comment]
@ssc327 
It looks like 5.x may need its own setting and fixes.  :sob: 

I don't have the bandwidth to make the changes, but you can use the PR I submitted as a template for adding the setting yourself. It is likely that another of the hooks being requested are not present in 5.x.  You'd need to look at the documentation to figure out which one(s).

**Comment by bitwiseman** [comment]
Looks like someone opened a new issue for 5.x, let's use that. #352

**Comment by dsakilesh** [comment]
Hello @bitwiseman @ssc327 

#352 has been closed.  Could you please let me know what is status of the **Native implementation** issue?

Thanks in advance!

**Comment by dsakilesh** [comment]
Hello @bitwiseman @ssc327 
I have updated our bitbucket version from **v5.8.2** to **v6.10.4** but we still have the same issue. Native implementation is not working. And this blocking us as we do not have access to **auto trigger PRs when target changes functionality.**

Kindly help us solve the issue. Thanks in advance!

**Comment by ssc327** [comment]
@dsakilesh after upgrading, have you tried native and a full scan from the top level folder?

**Comment by dsakilesh** [comment]
Hi @ssc327 I have tried it. but i don't see the PRs getting built automatically even though target branch commits are updated. Even the **Jenkins hook** is not created in Webhooks section in bitbucket repository settings.
Full scan will only trigger PRs when scan is trigerred but PRs should get triggered automatically whenever a target branch is updated for the PRs.

**Comment by ssc327** [comment]
@dsakilesh is your issue that webhooks are not created or that jobs for PRs do not get automatically triggered?

**Comment by dsakilesh** [comment]
@ssc327 My issue primarily is PRs do not get automatically triggered. And secondarily native implementation is not working as well (webhooks not automatically created by bitbucket branch source plugin)

**Comment by yogeek** [comment]
Hi all,

Today, I had the same issue with these versions :
```
Jenkins 2.235.5
Bitbucket Branch Source Plugin 2.9.2
Bitbucket Server v5.13.1
```

The webhook was not created anymore by Jenkins scan when I added a new repository in one of my existing bitbucket projects.

Then I changed jenkins user permissions in Bitbucket to make it "admin", and the webhook was created again when scanning a bitbucket organization !! but...
it was not configured correctly :( 
no events enables so no reaction to "push" events for example :
![image](https://user-images.githubusercontent.com/9500406/92502675-85fb9280-f200-11ea-80db-81bdc88c7f57.png)

So still a bug with this plugin compatibility with bitbucket v5.X apparently...
For now, users have to edit the webhook manually in bitbucket to make the push trigger jenkins job automatically for new repositories :(

**Comment by bitwiseman** [comment]
# IMPORTANT
If you are using BitBucket 5.x, go to #352.
